vec::IntoIter is now invariant

#35447 replaced a `*const T` by a `*mut T` in `IntoIter<T>`, which changed `IntoIter` from being covariant to being invariant. This breaks crates like `vec_map` which relied on `IntoIter` being invariant, e.g. https://travis-ci.org/asajeffrey/presort/jobs/152720189#L208 has error:
```
error[E0308]: mismatched types
   --> /home/ajeffrey/.cargo/registry/src/github.com-1ecc6299db9ec823/vec_map-0.6.0/src/lib.rs:952:84
    |
952 |     fn into_iter_covariant<'a, T>(iter: IntoIter<&'static T>) -> IntoIter<&'a T> { iter }
    |                                                                                    ^^^^ lifetime mismatch
    |
    = note: expected type `IntoIter<&'a T>`
    = note:    found type `IntoIter<&'static T>`
note: the lifetime 'a as defined on the block at 952:81...
   --> /home/ajeffrey/.cargo/registry/src/github.com-1ecc6299db9ec823/vec_map-0.6.0/src/lib.rs:952:82
    |
952 |     fn into_iter_covariant<'a, T>(iter: IntoIter<&'static T>) -> IntoIter<&'a T> { iter }
    |                                                                                  ^^^^^^^^
    = note: ...does not necessarily outlive the static lifetime

   Compiling rustc-serialize v0.3.19
error: aborting due to previous error
```
cc @alexcrichton @frewsxcv @apasel422

--- a/src/libcollections/vec.rs
+++ b/src/libcollections/vec.rs
@@ -1453,10 +1453,11 @@ impl<T> IntoIterator for Vec<T> {
             } else {
                 begin.offset(self.len() as isize) as *const T
             };
-            let buf = ptr::read(&self.buf);
+            let cap = self.buf.cap();
             mem::forget(self);
             IntoIter {
-                _buf: buf,
+                buf: Shared::new(begin),
+                cap: cap,
                 ptr: begin,
                 end: end,
             }
@@ -1708,8 +1709,9 @@ impl<'a, T> FromIterator<T> for Cow<'a, [T]> where T: Clone {
 /// [`IntoIterator`]: ../../std/iter/trait.IntoIterator.html
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct IntoIter<T> {
-    _buf: RawVec<T>,
-    ptr: *mut T,
+    buf: Shared<T>,
+    cap: usize,
+    ptr: *const T,
     end: *const T,
 }
 
@@ -1750,7 +1752,7 @@ impl<T> IntoIter<T> {
     #[unstable(feature = "vec_into_iter_as_slice", issue = "35601")]
     pub fn as_mut_slice(&self) -> &mut [T] {
         unsafe {
-            slice::from_raw_parts_mut(self.ptr, self.len())
+            slice::from_raw_parts_mut(self.ptr as *mut T, self.len())
         }
     }
 }
@@ -1846,9 +1848,10 @@ impl<T> Drop for IntoIter<T> {
     #[unsafe_destructor_blind_to_params]
     fn drop(&mut self) {
         // destroy the remaining elements
-        for _x in self {}
+        for _x in self.by_ref() {}
 
         // RawVec handles deallocation
+        let _ = unsafe { RawVec::from_raw_parts(*self.buf, self.cap) };
     }
 }
 

--- a/src/libcollectionstest/vec.rs
+++ b/src/libcollectionstest/vec.rs
@@ -11,7 +11,7 @@
 use std::borrow::Cow;
 use std::iter::{FromIterator, repeat};
 use std::mem::size_of;
-use std::vec::Drain;
+use std::vec::{Drain, IntoIter};
 
 use test::Bencher;
 
@@ -537,6 +537,7 @@ fn test_cow_from() {
 #[allow(dead_code)]
 fn assert_covariance() {
     fn drain<'new>(d: Drain<'static, &'static str>) -> Drain<'new, &'new str> { d }
+    fn into_iter<'new>(i: IntoIter<&'static str>) -> IntoIter<&'new str> { i }
 }
 
 #[bench]

